### ISSUE TITLE
ECAL fails when used in seperate CMake Project, TinyXML2 not properly packaged with cpack?

### ISSUE THREAD

**Issue by JeremyBYU** [continental/ecal#59]
I am attempting to use ECAL inside of a separate C++ project with CMake. I have been able to successfully do this with ECAL 5.5.2.  I am now trying to update and use 5.6.0. Inside my cloned `ecal` repo I have checked out the new release (tag) version and have updated all submodule dependencies.  Here is may CMake configuration when building `ecal` for completeness:

```
cmake .. -DCMAKE_BUILD_TYPE=Release -DECAL_THIRDPARTY_BUILD_PROTOBUF=OFF -DHAS_QT5=OFF -DBUILD_DOCS=OFF -DECAL_THIRDPARTY_BUILD_CURL=OFF -DECAL_THIRDPARTY_BUILD_HDF5=OFF -DBUILD_PY_BINDING=ON
make -j20
cpack -G DEB
sudo dpkg -i _deploy/eCAL-*
```

There are **no** errors building or installing ecal. However when attempting to now use Ecal in another cmake project (toally different directory) I get this error. Note the error is about not finding tinyxml target. This is a new dependency that it seems is not being bundled or found correctly when using `cpack`.
```
CMake Error at /usr/lib/cmake/eCAL/eCALTargets-release.cmake:9 (set_property):
  set_property could not find TARGET tinyxml2::tinyxml2.  Perhaps it has not
  yet been created.
Call Stack (most recent call first):
  /usr/lib/cmake/eCAL/eCALTargets.cmake:131 (include)
  /usr/lib/cmake/eCAL/eCALConfig.cmake:60 (include)
  CMakeLists.txt:33 (find_package)


CMake Error at /usr/lib/cmake/eCAL/eCALTargets-release.cmake:10 (set_target_properties):
  set_target_properties Can not find target to add properties to:
  tinyxml2::tinyxml2
Call Stack (most recent call first):
  /usr/lib/cmake/eCAL/eCALTargets.cmake:131 (include)
  /usr/lib/cmake/eCAL/eCALConfig.cmake:60 (include)
  CMakeLists.txt:33 (find_package)

```
My guess is that something has gone wrong with "packaging" ecal using cpack. Specifically the tinyxml2 dependency. Here's an image of this directory if needed. 

![image](https://user-images.githubusercontent.com/6628085/86967006-04bc4c80-c138-11ea-88b1-5e0ccc0e84fa.png)

Everything works perfectly fine and build if I use Ecal 5.5.2

Thanks for any help!

**Comment by rex-schilasky** [comment]
Hi Jeremy, first of all it's a good idea to try out the new 5.6 version. The improved recording concept is exciting (decentralized concept, user addon plugins, ftp upload ..). But first of all you should be able to link against it. What linux distribution are you using ?

We still did not found a perfect way to handle thirdparty dependencies for eCAL and to deploy them properly. That is an ongoing process. TinyXML is integrated as a submodule. You could try to install the package yourself in your system (https://www.howtoinstall.me/ubuntu/18-04/libtinyxml2-6/ for Ubuntu 18) and set the CMake option

```ECAL_THIRDPARTY_BUILD_TINYXML2 = OFF```

Then it should work fine hopefully.

**Comment by JeremyBYU** [comment]
I ran into some errors when trying to compile with the system `libtinyxml2-dev`. First `ecal` with `cmake` was unable to find the system package.  Ubuntu installs the header file and library in these locations:
```
(python38) ➜  realsense-tracking git:(master) dpkg -L libtinyxml2-dev    
/.
/usr
/usr/include
/usr/include/tinyxml2.h
/usr/lib
/usr/lib/x86_64-linux-gnu
/usr/lib/x86_64-linux-gnu/pkgconfig
/usr/lib/x86_64-linux-gnu/pkgconfig/tinyxml2.pc
/usr/share
/usr/share/doc
/usr/share/doc/libtinyxml2-dev
/usr/share/doc/libtinyxml2-dev/copyright
/usr/lib/x86_64-linux-gnu/libtinyxml2.so
/usr/share/doc/libtinyxml2-dev/changelog.Debian.gz

```
My system, with cmake, seems to be able to find libraries just fine when they are in `/usr/local`. However tinyxml2 did not install there. Therefore I went online and found this cmake file that should help inform cmake where to find tinyxml2. This file is taken from ROS, `Findtinyxml2.cmake`, placed at `cmake/Modules/Findtinyxml2.cmake`:

```cmake
# TinyXML2_FOUND
# TinyXML2_INCLUDE_DIRS
# TinyXML2_LIBRARIES

# try to find the CMake config file for TinyXML2 first
find_package(TinyXML2 CONFIG QUIET)
if(TinyXML2_FOUND)
  message(STATUS "Found TinyXML2 via Config file: ${TinyXML2_DIR}")
  if(NOT TINYXML2_LIBRARY)
    # in this case, we're probably using TinyXML2 version 5.0.0 or greater
    # in which case tinyxml2 is an exported target and we should use that
    if(TARGET tinyxml2)
      set(TINYXML2_LIBRARY tinyxml2)
    elseif(TARGET tinyxml2::tinyxml2)
      set(TINYXML2_LIBRARY tinyxml2::tinyxml2)
    elseif(TinyXML2_FIND_REQUIRED)
      message(FATAL_ERROR "Unable to determine target for TinyXML2")
    endif()
    list(APPEND TinyXML2_TARGETS ${TINYXML2_LIBRARY})
  else()
    # Only perform that logic once
    if(NOT TARGET tinyxml2::tinyxml2)
      # TINYXML2_LIBRARY is composed of debug;<path\to\debug.lib>;optimized;<path\to\release.lib>
      # we have to extract the appropriate component based on the current configuration.
      list(LENGTH TINYXML2_LIBRARY TINYXML_LIBRARY_LIST_LENGTH)
      if(NOT ${TINYXML_LIBRARY_LIST_LENGTH} EQUAL 4)
        message(FATAL_ERROR "Unable to extract the library file path from ${TINYXML2_LIBRARY}")
      endif()
      if(CMAKE_BUILD_TYPE MATCHES DEBUG)
        list(GET TINYXML2_LIBRARY 0 ASSERT_DEBUG)
        if(NOT ${ASSERT_DEBUG} STREQUAL "debug")
          message(FATAL_ERROR "could not parse debug library path from ${TINYXML2_LIBRARY}")
        endif()
        list(GET TINYXML2_LIBRARY 1 TINYXML2_LIBRARY_PATH)
      else()
        list(GET TINYXML2_LIBRARY 2 ASSERT_OPTIMIZED)
        if(NOT ${ASSERT_OPTIMIZED} STREQUAL "optimized")
          message(FATAL_ERROR "could not parse library path from ${TINYXML2_LIBRARY}")
        endif()
        list(GET TINYXML2_LIBRARY 3 TINYXML2_LIBRARY_PATH)
      endif()
      if(NOT EXISTS ${TINYXML2_LIBRARY_PATH})
        message(FATAL_ERROR "library file path ${TINYXML2_LIBRARY_PATH} does not exist")
      endif()
      
      add_library(tinyxml2::tinyxml2 UNKNOWN IMPORTED)
      set_property(TARGET tinyxml2::tinyxml2 PROPERTY IMPORTED_LOCATION ${TINYXML2_LIBRARY_PATH})
      set_property(TARGET tinyxml2::tinyxml2 PROPERTY INTERFACE_INCLUDE_DIRECTORIES ${TINYXML2_INCLUDE_DIR})
      list(APPEND TinyXML2_TARGETS tinyxml2::tinyxml2)
    endif()
  endif()
else()
  find_path(TINYXML2_INCLUDE_DIR NAMES tinyxml2.h)

  find_library(TINYXML2_LIBRARY tinyxml2)

  include(FindPackageHandleStandardArgs)
  find_package_handle_standard_args(TinyXML2 DEFAULT_MSG TINYXML2_LIBRARY TINYXML2_INCLUDE_DIR)

  mark_as_advanced(TINYXML2_INCLUDE_DIR TINYXML2_LIBRARY)

  if(NOT TARGET tinyxml2::tinyxml2)
    add_library(tinyxml2::tinyxml2 UNKNOWN IMPORTED)
    set_property(TARGET tinyxml2::tinyxml2 PROPERTY IMPORTED_LOCATION ${TINYXML2_LIBRARY})
    set_property(TARGET tinyxml2::tinyxml2 PROPERTY INTERFACE_INCLUDE_DIRECTORIES ${TINYXML2_INCLUDE_DIR})
    list(APPEND TinyXML2_TARGETS tinyxml2::tinyxml2)
  endif()
endif()

# Set mixed case INCLUDE_DIRS and LIBRARY variables from upper case ones.
if(NOT TinyXML2_INCLUDE_DIRS)
  set(TinyXML2_INCLUDE_DIRS ${TINYXML2_INCLUDE_DIR})
endif()
if(NOT TinyXML2_LIBRARIES)
  set(TinyXML2_LIBRARIES ${TINYXML2_LIBRARY})
endif()
```


Now when I run the following the build is configured correctly and tinyxml2 is found: `cmake .. -DCMAKE_BUILD_TYPE=Release -DECAL_THIRDPARTY_BUILD_PROTOBUF=OFF -DHAS_QT5=OFF -DBUILD_DOCS=OFF -DECAL_THIRDPARTY_BUILD_CURL=OFF -DECAL_THIRDPARTY_BUILD_HDF5=OFF -DBUILD_PY_BINDING=ON -DECAL_THIRDPARTY_BUILD_TINYXML2=OFF
`

However when I `make` a compile error occurs:

Error:
```txt
/home/jeremy/Documents/UMICH/Research/ecal/app/rec/rec_server_core/src/config/config_v2to3.cpp: In function ‘bool eCAL::rec_server::config_v2to3::writeConfigFile(const eCAL::rec_server::RecServerImpl&, const string&)’:
/home/jeremy/Documents/UMICH/Research/ecal/app/rec/rec_server_core/src/config/config_v2to3.cpp:205:125: error: ‘class tinyxml2::XMLDocument’ has no member named ‘GetErrorStr1’; did you mean ‘ErrorStr’?
             eCAL::rec::EcalRecLogger::Instance()->error(std::string("Error saving config to \"") + path + "\": " + document.GetErrorStr1());
                                                                                                                             ^~~~~~~~~~~~
                                                                                                                             ErrorStr
/home/jeremy/Documents/UMICH/Research/ecal/app/rec/rec_server_core/src/config/config.cpp: In function ‘bool eCAL::rec_server::config::readConfigFromFile(eCAL::rec_server::RecServerImpl&, const string&, int*)’:
/home/jeremy/Documents/UMICH/Research/ecal/app/rec/rec_server_core/src/config/config.cpp:74:126: error: ‘class tinyxml2::XMLDocument’ has no member named ‘GetErrorStr1’; did you mean ‘ErrorStr’?
             eCAL::rec::EcalRecLogger::Instance()->error(std::string("Error loading file from \"") + path + "\": " + document.GetErrorStr1());

```

Googling led me to these major API changes in Tinyxml2: [Link](https://github.com/leethomason/tinyxml2/issues/635).

Version I have installed: 

```
(python38) ➜  build git:(fe575cb) ✗ apt-cache policy libtinyxml2-dev                                         
libtinyxml2-dev:
  Installed: 6.0.0+dfsg-1
  Candidate: 6.0.0+dfsg-1
  Version table:
 *** 6.0.0+dfsg-1 500
        500 http://us.archive.ubuntu.com/ubuntu bionic/universe amd64 Packages
        100 /var/lib/dpkg/status

```

Version of commit tinyxml2 @ bf15233 (taken from CMakeLists.txt for this commit) -
```
set(GENERIC_LIB_VERSION "8.0.0")
set(GENERIC_LIB_SOVERSION "8")
```

Seems strange. The comment in the above mentioned issue makes me think that `ecal` is using the older deprecated method `GetErrorStr1`. However `ecals` thirdparty commit is using an even _newer_ version of `tinyxml2` than ubuntu 18.04! Maybe they un-deprecated the function names?  These errors (from api rename) only occurred in two files

1. `app/rec/rec_server_core/src/config/config.cpp`
2. `app/rec/rec_server_core/src/config/config_v2to3.cpp`

I changed them to use the newer function  `ErrorStr`. Then everything built correctly. I then built the deb package and installed.

In my  separate project I can now use `find_package(eCAL REQUIRED)` and it configured and builds correctly.

So i guess I would make two suggestions for a possible patch:

1. Add the `findtinyxml2.cmake` file
2. Fix the use of the older deprecated API usage of `GetErrorStr1`


EDIT - 

Okay I see the issue. These macros are not defined in older versions:
```
 #if TINYXML2_MAJOR_VERSION >= 6
```
Now they are. I guess thats what the issue was referring to.

**Comment by hannemn** [comment]
Hello @JeremyBYU ! First of all, thank you very much for your investigation regarding tinyxml! We will fix that issue in the upcoming patch version of eCAL. Actually the user of eCAL should not get in contact with TinyXML, because for linking against the eCAL library this dependency is not necessary (only for building eCAL from source). Another way to work around the bug is to build the eCAL library with the option -DBUILD_APPS=OFF. Then the critical parts gets excluded.

**Comment by JeremyBYU** [comment]
Thanks for the suggestion. I will use that flag for the build I have that don't need the ECAL monitoring tool.

**Comment by rex-schilasky** [comment]
Fixed with #64
